instrumentation/grape/lib/opentelemetry/instrumentation/grape/event_handler.rb

```diff
@@ -0,0 +1,133 @@
+            token = OpenTelemetry::Context.attach(OpenTelemetry::Trace.context_with_span(span))
+
+            payload.merge!(__opentelemetry_span: span, __opentelemetry_ctx_token: token)
+          end
+
+          # Handles the end of the endpoint_run.grape event (the parent event), where the context is detached
+          def endpoint_run_finish(_name, _id, payload)
+            span = payload.delete(:__opentelemetry_span)
+            token = payload.delete(:__opentelemetry_ctx_token)
+            return unless span && token
+
+            handle_payload_exception(span, payload[:exception_object]) if payload[:exception_object]
+
+            span.finish
+            OpenTelemetry::Context.detach(token)
+          end
+
+          # Handles the endpoint_render.grape event
+          def endpoint_render(_name, start, _finish, _id, payload)
+            name = span_name(payload[:endpoint])
+            attributes = {
+              'component' => 'template',
+              'operation' => 'endpoint_render'
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_payload_exception(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          # Handles the endpoint_run_filters.grape events
+          def endpoint_run_filters(_name, start, finish, _id, payload)
+            filters = payload[:filters]
+            type = payload[:type]
+
+            # Prevent submitting empty filters
+            return if (!filters || filters.empty?) || !type || (finish - start).zero?
+
+            name = span_name(payload[:endpoint])
+            attributes = {
+              'component' => 'web',
+              'operation' => 'endpoint_run_filters',
+              'grape.filter.type' => type.to_s
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_payload_exception(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          # Handles the format_response.grape event
+          def format_response(_name, start, _finish, _id, payload)
+            endpoint = payload[:env]['api.endpoint']
+            name = span_name(endpoint)
+            attributes = {
+              'component' => 'template',
+              'operation' => 'format_response',
+              'grape.formatter.type' => formatter_type(payload[:formatter])
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_payload_exception(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          private
+
+          def tracer
+            Grape::Instrumentation.instance.tracer
+          end
+
+          def span_name(endpoint)
```

Comment:
According to the [DDog porting guidelines](https://github.com/catawiki/opentelemetry-ruby-contrib/blob/55b6be133bd038fc597d8a08f7da661734eb0d38/instrumentation/datadog-porting-guide.md#L49), `resource` in DDog is equivalent to `span.name` here.

`resource` in DDog has the following structure: `BasicAPI GET /hello`, which includes the API class name, the method and the route. We decided to remove the API class name because we felt it didn't add much to the name (it's still in the attributes) and without this part it follows the recommended practice for http spans, as described [here](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/http/#name).

Also, I understand that we should keep cardinality low, and this is bound to have a relatively high cardinality on services with multiple endpoints, but if we went for something like 'GET' then it would be harder to group traces, so I'm wondering how low it should be in this case 🤔 And also, maybe a lower cardinality is needed for other instrumentations, but not for Grape? These are all open questions to us, so we'd be really grateful if you could provide more input and I'll be happy to make the necessary adjustments to the span name 🙏 😄

Reply:
Yes, the URLs are templated, so I'm leaving the span name as it is then 😄